moveit_kinematics/test/test_kinematics_plugin.cpp

```diff
@@ -0,0 +1,737 @@
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Jorge Nicho, Robert Haschke */
+
+#include <gtest/gtest.h>
+#include <memory>
+#include <boost/thread/mutex.hpp>
+#include <pluginlib/class_loader.h>
+#include <ros/ros.h>
+#include <tf2_eigen/tf2_eigen.h>
+#include <xmlrpcpp/XmlRpcValue.h>
+
+// MoveIt!
+#include <moveit/kinematics_base/kinematics_base.h>
+#include <moveit/rdf_loader/rdf_loader.h>
+#include <moveit/robot_model/robot_model.h>
+#include <moveit/robot_state/robot_state.h>
+
+#include <moveit/robot_state/conversions.h>
+#include <moveit_msgs/DisplayTrajectory.h>
+#include <moveit/robot_trajectory/robot_trajectory.h>
+
+const double IK_NEAR = 1e-5;
+
+const std::string ROBOT_DESCRIPTION_PARAM = "robot_description";
+const double DEFAULT_SEARCH_DISCRETIZATION = 0.01f;
+const double EXPECTED_SUCCESS_RATE = 0.8;
+
+template <typename T>
+inline bool getParam(const std::string& param, T& val)
+{
+  // first look within private namespace
+  ros::NodeHandle pnh("~");
+  if (pnh.getParam(param, val))
+    return true;
+
+  // then in local namespace
+  ros::NodeHandle nh;
+  if (nh.getParam(param, val))
+    return true;
+
+  return false;
+}
+
+// As loading of parameters is quite slow, we share them across all tests
+class SharedData
+{
+  friend class KinematicsTest;
+  typedef pluginlib::ClassLoader<kinematics::KinematicsBase> KinematicsLoader;
+
+  robot_model::RobotModelPtr robot_model_;
+  std::unique_ptr<KinematicsLoader> kinematics_loader_;
+  std::string root_link_;
+  std::string tip_link_;
+  std::string group_name_;
+  std::vector<std::string> joints_;
+  std::vector<double> seed_;
+  std::vector<double> consistency_limits_;
+  double timeout_;
+  int num_fk_tests_;
+  int num_ik_cb_tests_;
+  int num_ik_tests_;
+  int num_ik_multiple_tests_;
+  int num_nearest_ik_tests_;
+
+  SharedData(SharedData const&) = delete;  // this is a singleton
+  SharedData()
+  {
+    initialize();
+  }
+
+  void initialize()
+  {
+    ROS_INFO_STREAM("Loading robot model from " << ros::this_node::getNamespace() << "/" << ROBOT_DESCRIPTION_PARAM);
+    // load robot model
+    rdf_loader::RDFLoader rdf_loader(ROBOT_DESCRIPTION_PARAM);
+    robot_model_ = std::make_shared<robot_model::RobotModel>(rdf_loader.getURDF(), rdf_loader.getSRDF());
```

Comment:
Let's use our growing testing framework for MoveIt!, see 

>     robot_model_ = moveit::core::loadTestingRobotModel("panda_description");

Documented in tutorials: https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html

@BryceStevenWilley

Reply:
I looked into this possibility before, but it doesn't work here because the test utils load the robot models directly from urdf / srdf files in `moveit_resources`. However, here I load them from the parameter server.